homeassistant/components/fireservicerota/__init__.py

```diff
@@ -0,0 +1,182 @@
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up FireServiceRota from a config entry."""
+    coordinator = FSRDataUpdateCoordinator(hass, entry)
+    await coordinator.async_update()
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    for platform in SUPPORTED_PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload FireServiceRota config entry."""
+    hass.data[DOMAIN][entry.entry_id].stop_ws_listener()
+
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+                for platform in SUPPORTED_PLATFORMS
+            ]
+        )
+    )
+
+    if unload_ok:
+        del hass.data[DOMAIN][entry.entry_id]
+
+    return unload_ok
+
+
+class FSRDataUpdateCoordinator:
+    """Getting the latest data from fireservicerota."""
+
+    def __init__(self, hass, entry):
+        """Initialize the data object."""
+        self._hass = hass
+        self._entry = entry
+
+        self._url = entry.data[CONF_URL]
+        self._tokens = entry.data[CONF_TOKEN]
+
+        self.availability_data = None
+        self.incident_data = None
+        self.incident_id = None
+        self.response_data = None
+
+        self.fsr_avail = FireServiceRota(
+            base_url=f"https://{self._url}", token_info=self._tokens
+        )
+
+        self.fsr_incidents = FireServiceRotaIncidents(on_incident=self.on_incident)
+
+        self.start_ws_listener()
+
+    def construct_url(self) -> str:
+        """Return url with latest config values."""
+        return WSS_BWRURL.format(
+            self._entry.data[CONF_URL], self._entry.data[CONF_TOKEN]["access_token"]
+        )
+
+    def on_incident(self, data) -> None:
+        """Update the current data."""
+        _LOGGER.debug("Got data from websocket listener: %s", data)
+        self.incident_data = data
+
+        async_dispatcher_send(self._hass, f"{DOMAIN}_{self._entry.entry_id}_update")
+
+    def start_ws_listener(self) -> None:
+        """Start the websocket listener."""
+        _LOGGER.debug("Starting incidents listener")
+        self.fsr_incidents.start(self.construct_url())
+
+    def stop_ws_listener(self) -> None:
+        """Stop the websocket listener."""
+        _LOGGER.debug("Stopping incidents listener")
+        self.fsr_incidents.stop()
+
+    async def update_call(self, func, *args):
+        """Perform update call and return data."""
+        try:
+            return await self._hass.async_add_executor_job(func, *args)
+        except (ExpiredTokenError, InvalidTokenError):
+            if await self.async_refresh_tokens():
+                return await self._hass.async_add_executor_job(func, *args)
```

Comment:
Add a guard to only refresh once so you don't accidentally end up in a loop and DDOS their server.

Also raise the exception if you don't end up refreshing the tokens, or else you're going to return `None` and that will end up in `self.availability_data`. Or is that expected?

```suggestion
    async def update_call(self, func, *args, tokens_refreshed=False):
        """Perform update call and return data."""
        try:
            return await self._hass.async_add_executor_job(func, *args)
        except (ExpiredTokenError, InvalidTokenError):
            if not tokens_refreshed and await self.async_refresh_tokens():
                return await self._hass.async_add_executor_job(
                    func, *args, tokens_refreshed=True
                )
            raise
```

Reply:
I have added code which seems to work if I make refresh and access token invalid, websocket is stopped, DataUpdateCoordinator doesn't poll, the update_calls are safeguarded and a notification is raise for the user.
self.availability_data being None is not an issue, I check for this. (in binary_sensor, not yet included)